### ISSUE TITLE
I tried to deploy locally in MAC OSX, but it fails

### ISSUE THREAD

**Issue by NAlhirabi** [mike-goodwin/owasp-threat-dragon#116]
Hello Mike 
I tried to deploy locally in MAC OSX, but it fails , when I run
npm start
Terminal shows following error:

> owasp-threat-dragon@0.6.1 start /Users/nadaalhirabi/Documents/GitHub/owasp-threat-dragon
> node server.js

{"name":"threatdragon","excludes":["req-headers","res-headers","res","req","short-body","body","response-hrtime","incoming","user-agent","response-time","http-version"],"hostname":"m006.cs.cf.ac.uk","pid":81568,"level":50,"security":true,"msg":"secure session cookie flag was false - should only happen in dev environments","time":"2019-11-11T14:07:32.065Z","v":0}
{"name":"threatdragon","hostname":"m006.cs.cf.ac.uk","pid":81568,"level":50,"msg":"owasp threat dragon failed to start up","time":"2019-11-11T14:07:32.067Z","v":0}
{"name":"threatdragon","hostname":"m006.cs.cf.ac.uk","pid":81568,"level":50,"msg":"Credentials must be provided when creating a service client.","time":"2019-11-11T14:07:32.067Z","v":0}
events.js:187
      throw er; // Unhandled 'error' event
      ^

Error: listen EADDRINUSE: address already in use :::3000
    at Server.setupListenHandle [as _listen2] (net.js:1300:14)
    at listenInCluster (net.js:1348:12)
    at Server.listen (net.js:1436:7)
    at Function.listen (/Users/nadaalhirabi/Documents/GitHub/owasp-threat-dragon/node_modules/express/lib/application.js:618:24)
    at Object.<anonymous> (/Users/nadaalhirabi/Documents/GitHub/owasp-threat-dragon/server.js:7:18)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Function.Module.runMain (internal/modules/cjs/loader.js:1025:10)
Emitted 'error' event on Server instance at:
    at emitErrorNT (net.js:1327:8)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) {
  code: 'EADDRINUSE',
  errno: 'EADDRINUSE',
  syscall: 'listen',
  address: '::',
  port: 3000
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! owasp-threat-dragon@0.6.1 start: `node server.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the owasp-threat-dragon@0.6.1 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/nadaalhirabi/.npm/_logs/2019-11-11T14_07_32_140Z-debug.log
m006:owasp-threat-dragon nadaalhirabi$ 

————

This is the variables in bash_profile

GITHUB_CLIENT_ID=XXXXXX
GITHUB_CLIENT_SECRET=XXXXXX
SESSION_SIGNING_KEY=XXXXX (note it is the Personal access tokens in Github)
SESSION_STORE=local
AZURE_STORAGE_CONNECTION_STRING
NODE_ENV=development

**Comment by jgadsden** [comment]
Hello @NAlhirabi - this is probably due to a previous process still using this address. Have you got a crashed ThreatDragon or another application running that is still listening on 3000? Can you try `ps -eax |grep node` to see what node processes are running - you can use `kill -9 <PID>` on the PID if there are any unexpected Threat Dragon instances.

**Comment by NAlhirabi** [comment]
thats what happened: and in the browser when I put (http://localhost:3000 ) the page shows Cannot GET /





m006:owasp-threat-dragon nadaalhirabi$ ps -eax | grep node
72589 ??         0:43.88 /private/var/folders/7n/czs033q94f1bvpj_9zsnvyg00000gn/T/AppTranslocation/B1130A94-5378-4331-AF06-E13708D28D6D/d/GitHub Desktop.app/Contents/Frameworks/GitHub Desktop Helper.app/Contents/MacOS/GitHub Desktop Helper --type=renderer --disable-features=SpareRendererForSitePerProcess --service-pipe-token=344345926303557290 --lang=en-GB --app-path=/private/var/folders/7n/czs033q94f1bvpj_9zsnvyg00000gn/T/AppTranslocation/B1130A94-5378-4331-AF06-E13708D28D6D/d/GitHub Desktop.app/Contents/Resources/app --enable-experimental-web-platform-features --node-integration --no-sandbox --no-zygote --background-color=#fff --disable-blink-features=Auxclick --num-raster-threads=3 --enable-zero-copy --enable-gpu-memory-buffer-compositor-resources --enable-main-frame-before-activation --service-request-channel-token=344345926303557290 --renderer-client-id=4 --no-v8-untrusted-code-mitigations
75954 ??         0:00.58 /Applications/Microsoft Teams.app/Contents/Frameworks/Microsoft Teams Helper.app/Contents/MacOS/Microsoft Teams Helper --type=renderer --autoplay-policy=no-user-gesture-required --service-pipe-token=A92E20E8E21C8DC37046081322F409FA --lang=en-GB --app-path=/Applications/Microsoft Teams.app/Contents/Resources/app.asar --node-integration=false --webview-tag=false --enable-sandbox --preload=/Applications/Microsoft Teams.app/Contents/Resources/app.asar/lib/renderer/notifications/preload_notifications_sandbox.js --background-color=#fff --num-raster-threads=3 --enable-zero-copy --enable-gpu-memory-buffer-compositor-resources --enable-main-frame-before-activation --enable-compositor-image-animations --service-request-channel-token=A92E20E8E21C8DC37046081322F409FA --renderer-client-id=7 --msteams-process-type=notificationsManager
81897 ??         0:32.31 /Applications/Microsoft Teams.app/Contents/Frameworks/Microsoft Teams Helper.app/Contents/MacOS/Microsoft Teams Helper --type=renderer --autoplay-policy=no-user-gesture-required --service-pipe-token=CCF1D9D25B29E58404835D5EF23D9E96 --lang=en-GB --app-path=/Applications/Microsoft Teams.app/Contents/Resources/app.asar --node-integration=false --webview-tag=true --no-sandbox --preload=/Applications/Microsoft Teams.app/Contents/Resources/app.asar/lib/renderer/preload.js --background-color=#fff --num-raster-threads=3 --enable-zero-copy --enable-gpu-memory-buffer-compositor-resources --enable-main-frame-before-activation --enable-compositor-image-animations --service-request-channel-token=CCF1D9D25B29E58404835D5EF23D9E96 --renderer-client-id=14 --msteams-process-type=mainWindow
81899 ??         0:00.58 /Applications/Microsoft Teams.app/Contents/Frameworks/Microsoft Teams Helper.app/Contents/MacOS/Microsoft Teams Helper --type=renderer --autoplay-policy=no-user-gesture-required --service-pipe-token=0831A560834F8500D7AA5B46DB5CA115 --lang=en-GB --app-path=/Applications/Microsoft Teams.app/Contents/Resources/app.asar --node-integration=false --webview-tag=true --enable-sandbox --native-window-open --preload=/Applications/Microsoft Teams.app/Contents/Resources/app.asar/lib/renderer/webview/webview-preload.js --background-color=#fff --guest-instance-id=2 --enable-blink-features --disable-blink-features --hidden-page --num-raster-threads=3 --enable-zero-copy --enable-gpu-memory-buffer-compositor-resources --enable-main-frame-before-activation --enable-compositor-image-animations --service-request-channel-token=0831A560834F8500D7AA5B46DB5CA115 --renderer-client-id=17
81900 ??         0:03.53 /Applications/Microsoft Teams.app/Contents/Frameworks/Microsoft Teams Helper.app/Contents/MacOS/Microsoft Teams Helper --type=renderer --autoplay-policy=no-user-gesture-required --service-pipe-token=78827C2B2B41C2987CBA6C564145A2B6 --lang=en-GB --app-path=/Applications/Microsoft Teams.app/Contents/Resources/app.asar --node-integration=false --webview-tag=false --no-sandbox --preload=/Applications/Microsoft Teams.app/Contents/Resources/app.asar/lib/pluginhost/preload.js --background-color=#fff --num-raster-threads=3 --enable-zero-copy --enable-gpu-memory-buffer-compositor-resources --enable-main-frame-before-activation --enable-compositor-image-animations --service-request-channel-token=78827C2B2B41C2987CBA6C564145A2B6 --renderer-client-id=19 --msteams-process-type=pluginHost
81547 ttys002    0:00.54 node server.js
82399 ttys002    0:00.00 grep node
m006:owasp-threat-dragon nadaalhirabi$ kill -9 <81547>
-bash: syntax error near unexpected token `81547'
m006:owasp-threat-dragon nadaalhirabi$ kill -9 81547
m006:owasp-threat-dragon nadaalhirabi$ kill -9 82399
-bash: kill: (82399) - No such process
m006:owasp-threat-dragon nadaalhirabi$ npm start

> owasp-threat-dragon@0.6.1 start /Users/nadaalhirabi/.Trash/owasp-threat-dragon
> node server.js

{"name":"threatdragon","excludes":["req-headers","res-headers","res","req","short-body","body","response-hrtime","incoming","user-agent","response-time","http-version"],"hostname":"m006.cs.cf.ac.uk","pid":82424,"level":50,"security":true,"msg":"secure session cookie flag was false - should only happen in dev environments","time":"2019-11-12T10:37:46.574Z","v":0}
{"name":"threatdragon","hostname":"m006.cs.cf.ac.uk","pid":82424,"level":50,"msg":"owasp threat dragon failed to start up","time":"2019-11-12T10:37:46.576Z","v":0}
{"name":"threatdragon","hostname":"m006.cs.cf.ac.uk","pid":82424,"level":50,"msg":"Credentials must be provided when creating a service client.","time":"2019-11-12T10:37:46.576Z","v":0}

**Comment by jgadsden** [comment]
Hello @NAlhirabi - I guess that there is something wrong in the environment values for your github account; Client ID and Client Secret and session signing key.
Do you want to try the desktop application first? It is easier to set up and once you have evaluated it you can then move on to using github to store your model files.
Threat Dragon Desktop version is here https://github.com/mike-goodwin/owasp-threat-dragon-desktop
and the install is easy once you have downloaded the source:
https://github.com/mike-goodwin/owasp-threat-dragon-desktop/archive/master.zip
or cloned it:
`git clone https://github.com/mike-goodwin/owasp-threat-dragon-desktop`
`npm install`
`npm run start`

**Comment by NAlhirabi** [comment]
Thank you Jon.
The desktop version works 😊
I have been trying to find the code file that enable me to add some shapes / edit the available one but I could not ( I would like to do some changes and see if there is ability to add some features to threat dragon , I am learning )
In the web version it was in td>>public>>app>> threatdragon.js folder but in the desktop version I do not know where to find the folder that let me add more shapes!!
Do I need to download something !!
Wher should I look at.

Thanks again for your suggestion.
Regards, Nada.

**Comment by jgadsden** [comment]
Hello @NAlhirabi  - the good news is that this functionality is in the core files that are common to both desktop and web Threat Dragon. These core files are held in a third repo https://github.com/mike-goodwin/owasp-threat-dragon-core. I am not sure exactly what you are trying to change, but the changes should be applied in the core files which are then used to create the mammoth threatdragon.js file.
If you are protoyping then you may want to look at the files in your node-modules directory, such as `node_modules/owasp-threat-dragon-core/src/diagrams/` and apply changes there?
Hope this helps,
Cheers,
Jon

**Comment by hardword** [comment]
Hi @NAlhirabi 

I had the same problem and spent some time to debug. It finally works with my OSX env. and I'd like to share my experience.

In my machine:
`export GITHUB_CLIENT_ID=XXXXXX`
`export GITHUB_CLIENT_SECRET=XXXXXX`
`export SESSION_SIGNING_KEY=XXXX(32 byte long random key)` <-- diff from your setting
`export SESSION_ENCRYPTION_KEYS='[{"isPrimary": true, "id": 0, "value": "XXXX(32 byte)"}, {"isPrimary": false, "id": 1, "value": "XXXX(32 byte)"}]'`
`export SESSION_STORE=local`
`export NODE_ENV=development`

In Github (the same as yours)
Homepage URL: http://localhost:3000
Authorization callback URL: http://localhost:3000/oauth/github

It seems
- I needed to explicitly "export" the (env) variables. (Local var assigning does not seem to work)
- I also needed to export "SESSION_ENCRYPTION_KEYS" var.
- And the key size should be 32 byte. (I found this through debug message but was not clear if the error is for SESSION_SIGNING_KEY or SESSION_ENCRYPTION_KEYS. So.. I just give 32byte keys to both.)

Hope this helps
